ckan/views/__init__.py

```diff
@@ -43,25 +46,93 @@ def set_cors_headers_for_response(response: Response) -> Response:
     return response
 
 
+def set_etag_and_fast_304_response_if_unchanged(response: Response) -> Response:
+    """Set ETag and return 304 if content is unchanged."""
+
+    enable_etags = config.get(u'ckan.cache_etags', True)
+    enable_etags_not_modified = config.get(u'ckan.cache_etags_notModified', True)
+    allowed_status_codes = {HTTPStatus.OK,  # 200
+                            HTTPStatus.MOVED_PERMANENTLY,  # 301
+                            HTTPStatus.FOUND,  # 302
+                            HTTPStatus.UNAUTHORIZED,  # 401
+                            HTTPStatus.FORBIDDEN,  # 403
+                            HTTPStatus.NOT_FOUND  # 404
+                            }
+    if response.status_code in allowed_status_codes and enable_etags:
+        if 'etag' not in response.headers:
+            # s3 etag uses md5, so using it here also
+
+            content_type = response.mimetype or ''
+            allowed_types = {'text/plain', 'text/css', 'text/html', 'application/json'}
+
+            try:
+                if content_type not in allowed_types:
+                    data_to_hash = response.get_data()
```

Comment:
IIUC this will break any streaming responses e.g. `/datastore/dump/<resource_id>`

Reply:
good to know, I'll get that addressed.